c/rsa_sighash_all.c

```diff
@@ -145,3 +180,209 @@ int md_string(const mbedtls_md_info_t *md_info, const unsigned char *buf,
   mbedtls_md_free(&ctx);
   return ret;
 }
+
+/*
+ * The following code is to add RSA "validate all" method.
+ * It mimic the behavior of validate_secp256k1_blake2b_sighash_all.
+ */
+#ifdef CKB_USE_SIM
+#include "ckb_consts.h"
+#include "ckb_syscall_sim.h"
+#else
+#include "ckb_syscalls.h"
+#endif
+#include "blake2b.h"
+#include "blockchain.h"
+
+#define BLAKE2B_BLOCK_SIZE 32
+#define BLAKE160_SIZE 20
+/* 32 KB */
+#define MAX_WITNESS_SIZE 32768
+#define SCRIPT_SIZE 32768
+#define TEMP_SIZE 32768
+#define ONE_BATCH_SIZE 32768
+
+int load_and_hash_witness(blake2b_state *ctx, size_t index, size_t source) {
+  uint8_t temp[ONE_BATCH_SIZE];
+  uint64_t len = ONE_BATCH_SIZE;
+  int ret = ckb_load_witness(temp, &len, 0, index, source);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  blake2b_update(ctx, (char *)&len, sizeof(uint64_t));
+  uint64_t offset = (len > ONE_BATCH_SIZE) ? ONE_BATCH_SIZE : len;
+  blake2b_update(ctx, temp, offset);
+  while (offset < len) {
+    uint64_t current_len = ONE_BATCH_SIZE;
+    ret = ckb_load_witness(temp, &current_len, offset, index, source);
+    if (ret != CKB_SUCCESS) {
+      return ret;
+    }
+    uint64_t current_read =
+        (current_len > ONE_BATCH_SIZE) ? ONE_BATCH_SIZE : current_len;
+    blake2b_update(ctx, temp, current_read);
+    offset += current_read;
+  }
+  return CKB_SUCCESS;
+}
+
+// Extract lock from WitnessArgs
+int extract_witness_lock(uint8_t *witness, uint64_t len,
+                         mol_seg_t *lock_bytes_seg) {
+  mol_seg_t witness_seg;
+  witness_seg.ptr = witness;
+  witness_seg.size = len;
+
+  if (MolReader_WitnessArgs_verify(&witness_seg, false) != MOL_OK) {
+    return ERROR_ENCODING;
+  }
+  mol_seg_t lock_seg = MolReader_WitnessArgs_get_lock(&witness_seg);
+
+  if (MolReader_BytesOpt_is_none(&lock_seg)) {
+    return ERROR_ENCODING;
+  }
+  *lock_bytes_seg = MolReader_Bytes_raw_bytes(&lock_seg);
+  return CKB_SUCCESS;
+}
+
+int load_public_key_hash(unsigned char *public_key) {
+  int ret;
+  uint64_t len = 0;
+
+  /* Load args */
+  unsigned char script[SCRIPT_SIZE];
+  len = SCRIPT_SIZE;
+  ret = ckb_load_script(script, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+  if (len > SCRIPT_SIZE) {
+    return ERROR_SCRIPT_TOO_LONG;
+  }
+  mol_seg_t script_seg;
+  script_seg.ptr = (uint8_t *)script;
+  script_seg.size = len;
+
+  if (MolReader_Script_verify(&script_seg, false) != MOL_OK) {
+    return ERROR_ENCODING;
+  }
+
+  mol_seg_t args_seg = MolReader_Script_get_args(&script_seg);
+  mol_seg_t args_bytes_seg = MolReader_Bytes_raw_bytes(&args_seg);
+  if (args_bytes_seg.size != PUBLIC_KEY_SIZE1) {
+    return ERROR_WRONG_SCRIPT_ARGS_LEN;
+  }
+  memcpy(public_key, args_bytes_seg.ptr, args_bytes_seg.size);
+  return CKB_SUCCESS;
+}
+
+// this method performs RSA signature verification: 1024-bits.
```

Comment:
Is there a way we can somehow detect key bits from public key information? Locking here to 1024 bits sounds like too big a restriction to me.

Reply:
It's key size independent. I will check all the comments.